### PR TITLE
[DET-3186] Avoid crashing when using IndexedSlices in TF2.

### PR DESCRIPTION
**Pull request by aaron276h**
Because we execute our gradient aggregation code in graph mode, IndexedSlices
crash when we divide aggregation_frequency (this code is never executed, only
built because of graph mode).

### CODE REVIEW

File: horovod/common/gradient_aggregation_eager.py [hunk 1]
```diff
@@ -74,12 +74,17 @@ def compute_gradients(self, grads):
...
+                gradient_divisor = (
+                    self.aggregation_frequency if self.average_aggregated_gradients else 1
+                )
+                allreduced_grads = [
+                    grad / gradient_divisor if grad is not None else None for grad in grads
+                ]
+            except Exception as e:
+                print(f"Warning: exception {e} generated during aggregation division. If you are using "
+                      "Indexed slices this is expected. If you are using aggregation of 1 this exception"
+                      "will have no effect.")
```
Comment (brain-good) [inline horovod/common/gradient_aggregation_eager.py:87]: >  Indexed slices this is expected. If you are using aggregation of 1 this exception will have no effect.

question: this seems a little odd? Can we instead gate this with some if statements and not catch exceptions?

Comment (aaron276h) [reply horovod/common/gradient_aggregation_eager.py]: Took a new approach on this, after being able to reproduce the original error.

File: horovod/common/gradient_aggregation_eager.py [hunk 2]
```diff
@@ -1,4 +1,25 @@
+import functools
+import os
 import tensorflow as tf
+from packaging import version
+
+
+def decorate_before_tf_2_2(func):
+    if version.parse(tf.__version__) < version.parse("2.2.0") \
+            or bool(int(os.getenv("ENABLE_TF_FUNCTION_FOR_ALLREDUCE", "0"))):
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/common/gradient_aggregation_eager.py]: Good call, update this.

File: horovod/common/gradient_aggregation_eager.py [hunk 3]
```diff
@@ -1,4 +1,25 @@
+import functools
+import os
 import tensorflow as tf
+from packaging import version
+
+
+def decorate_before_tf_2_2(func):
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/common/gradient_aggregation_eager.py]: Yeah true, i tweaked the name, lmk what you think of the new one.

File: horovod/common/gradient_aggregation_eager.py [hunk 4]
```diff
@@ -1,4 +1,26 @@
+import functools
+import os
 import tensorflow as tf
+from distutils import util
+from packaging import version
+
+
+def tf_function_before_tf_2_2(func):
+    if version.parse(tf.__version__) < version.parse("2.2.0") \
+            or util.strtobool(os.getenv("ENABLE_TF_FUNCTION_FOR_ALLREDUCE", "False")) == 1:
```
Comment (brain-good) [inline horovod/common/gradient_aggregation_eager.py:10]: We should remove the `== 1`

Comment (aaron276h) [reply horovod/common/gradient_aggregation_eager.py]: done

**Pull request merged**